tool/src/js.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::HashMap;
+
+use diplomat_core::ast;
+
+pub fn gen_bindings(env: &HashMap<String, ast::CustomType>) -> Vec<String> {
+    let mut out = vec!["import wasm from \"./wasm.mjs\"".to_string()];
+
+    for custom_type in env.values() {
+        gen_struct(&mut out, custom_type, env);
+    }
+
+    out
+}
+
+fn gen_struct(
+    out: &mut Vec<String>,
+    custom_type: &ast::CustomType,
+    env: &HashMap<String, ast::CustomType>,
+) {
+    out.push(format!(
+        "const {}_destroy_registry = new FinalizationRegistry(underlying => {{",
+        custom_type.name()
+    ));
+    out.push(format!("wasm.{}_destroy(underlying);", custom_type.name()));
+    out.push("});".to_string());
+
+    out.push(format!("export class {} {{", custom_type.name()));
+
+    out.push("constructor(underlying) {".to_string());
+    out.push("this.underlying = underlying;".to_string());
+    out.push("}".to_string());
+
+    for method in custom_type.methods().iter() {
+        gen_method(method, &env, out);
+    }
+
+    out.push("}".to_string());
+}
+
+fn gen_method(method: &ast::Method, env: &HashMap<String, ast::CustomType>, out: &mut Vec<String>) {
+    let all_params = method
+        .params
+        .iter()
+        .map(|p| p.name.clone())
+        .collect::<Vec<String>>()
+        .join(", ");
+    if method.self_param.is_some() {
+        out.push(format!("{}({}) {{", method.name, &all_params));
+        match &method.return_type {
+            Some(ret_type) => {
+                let value = gen_value(
+                    format!(
+                        "wasm.{}(this.underlying, {})",
+                        method.full_path_name, all_params
+                    ),
+                    ret_type,
+                    env,
+                );
+                out.push(format!("return {};", value));
+            }
+
+            None => {
+                out.push(format!(
+                    "wasm.{}(this.underlying, {});",
+                    method.full_path_name, all_params
+                ));
+            }
+        }
+    } else {
+        out.push(format!("static {}({}) {{", method.name, &all_params));
+        match &method.return_type {
+            Some(ret_type) => {
+                let value = gen_value(
+                    format!("wasm.{}({})", method.full_path_name, all_params),
+                    ret_type,
+                    env,
+                );
+                out.push(format!("return {};", value));
+            }
+
+            None => {
+                out.push(format!("wasm.{}({});", method.full_path_name, all_params));
+            }
+        }
+    }
+
+    out.push("}".to_string());
+}
+
+fn gen_value(
```

Comment:
I would call this something like `convert_value_to_js` or something so that it's clear which way it's going in? I'm pretty sure for struct input types we're going to need the reverse function too.